File: ruby/lib/action.rb
# frozen_string_literal: true

# Actino is a block of logic with a human readable name
class Action
  attr_reader :name

  def initialize(name, &callback)
    @name = name
    @callback = callback
  end

  def do
    @callback&.call(self)
  end

  def to_s
    @name
  end
end

def Action(name, &callback)
  Action.new(name, &callback)
end

File: golang/entities/action.go
package entities

type ActionFunc func(*Action)

type Action struct {
	Name     string
	callback ActionFunc
}

func NewAction(name string, callback ActionFunc) *Action {
	return &Action{Name: name, callback: callback}
}

func (a *Action) Do() {
	a.callback(a)
}

func (a *Action) String() string {
	return a.Name
}
